hail/src/main/scala/is/hail/types/physical/stypes/concrete/SSubsetStruct.scala

```diff
@@ -6,47 +6,59 @@ import is.hail.expr.ir.{EmitCodeBuilder, IEmitCode}
+import is.hail.types.virtual.{Field, TStruct, Type}
 
-final case class SSubsetStruct(parent: SBaseStruct, fieldNames: IndexedSeq[String])
-    extends SBaseStruct {
+class SSubsetStruct(
+  private val parent: SBaseStruct,
+  private val fieldNames: IndexedSeq[String],
+) extends SBaseStruct {
 
   override val size: Int = fieldNames.size
 
   val _fieldIdx: Map[String, Int] = fieldNames.zipWithIndex.toMap
 
-  val newToOldFieldMapping: Map[Int, Int] = _fieldIdx
-    .map { case (f, i) => (i, parent.virtualType.asInstanceOf[TStruct].fieldIdx(f)) }
+  lazy val newToOldFieldMapping: Map[Int, Int] = {
+    val parentFieldIdx = parent.virtualType.asInstanceOf[TStruct]
+    _fieldIdx.map { case (f, i) => i -> parentFieldIdx.fieldIdx(f) }
+  }
 
-  override val fieldTypes: IndexedSeq[SType] =
+  override lazy val fieldTypes: IndexedSeq[SType] =
     Array.tabulate(size)(i => parent.fieldTypes(newToOldFieldMapping(i)))
 
-  override val fieldEmitTypes: IndexedSeq[EmitType] =
+  override lazy val fieldEmitTypes: IndexedSeq[EmitType] =
     Array.tabulate(size)(i => parent.fieldEmitTypes(newToOldFieldMapping(i)))
 
   override lazy val virtualType: TStruct = {
-    val vparent = parent.virtualType.asInstanceOf[TStruct]
-    TStruct(fieldNames.map(f => (f, vparent.field(f).typ)): _*)
+    val vparentTypes = parent.virtualType.asInstanceOf[TStruct].types
+    TStruct(fieldNames.zipWithIndex.map { case (f, i) =>
+      Field(f, vparentTypes(newToOldFieldMapping(i)), i)
+    })
   }
 
-  override def fieldIdx(fieldName: String): Int = _fieldIdx(fieldName)
+  override def fieldIdx(fieldName: String): Int =
+    _fieldIdx(fieldName)
 
   override def castRename(t: Type): SType = {
-    val renamedVType = t.asInstanceOf[TStruct]
-    val newNames = renamedVType.fieldNames
-    val subsetPrevVirtualType = virtualType
-    val vparent = parent.virtualType.asInstanceOf[TStruct]
-    val newParent = TStruct(vparent.fieldNames.map(f =>
-      subsetPrevVirtualType.fieldIdx.get(f) match {
-        case Some(idxInSelectedFields) =>
-          val renamed = renamedVType.fields(idxInSelectedFields)
-          (renamed.name, renamed.typ)
-        case None => (f, vparent.fieldType(f))
-      }
-    ): _*)
-    val newType = SSubsetStruct(parent.castRename(newParent).asInstanceOf[SBaseStruct], newNames)
-    assert(newType.virtualType == t)
-    newType
+    val newVirtualType = t.asInstanceOf[TStruct]
+    val oldToNewFieldMapping = newToOldFieldMapping.map(n => n._2 -> n._1)
+
+    // note we may have subsetted a parent struct{x,y,z} to struct{z} then renamed to struct{x}
+    // must only tell parent to castRename what it knows as `z`, leaving others intact
+    val newParent = parent.castRename(
+      TStruct(parent.virtualType.fields.zipWithIndex.map { case (f, i) =>
+        oldToNewFieldMapping.get(i) match {
+          case Some(idx) => f.copy(typ = newVirtualType.types(idx))
+          case None => f
+        }
+      })
+    )
```

Comment:
Do we need this? I thought the point of defining equality to ignore field names in the parent was to not have to rename the parent at all.

Reply:
No longer needed